File: tc-dotnet/KeyDerivation.cs
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using System.Security.Cryptography;

namespace TrueCryptReader;

/// <summary>
/// PBKDF2 key derivation supporting all TrueCrypt PRFs.
/// </summary>
public enum TrueCryptPrf
{
    RipeMd160 = 1,
    Sha512 = 2,
    Whirlpool = 3,
    Sha1 = 4,
}

public static class KeyDerivation
{
    /// <summary>All PRFs to try, in order.</summary>
    public static readonly TrueCryptPrf[] AllPrfs =
    [
        TrueCryptPrf.RipeMd160,
        TrueCryptPrf.Sha512,
        TrueCryptPrf.Whirlpool,
        TrueCryptPrf.Sha1,
    ];

    /// <summary>
    /// Returns the PBKDF2 iteration count for a given PRF, matching TrueCrypt's Pkcs5.c.
    /// </summary>
    public static int GetIterations(TrueCryptPrf prf, bool boot)
    {
        return prf switch
        {
            TrueCryptPrf.RipeMd160 => boot ? 1000 : 2000,
            TrueCryptPrf.Sha512 => 1000,
            TrueCryptPrf.Whirlpool => 1000,
            TrueCryptPrf.Sha1 => 2000,
            _ => throw new ArgumentException($"Unknown PRF: {prf}")
        };
    }

    /// <summary>
    /// Derives a key using PBKDF2-HMAC with the specified PRF.
    /// Output is 256 bytes (MASTER_KEYDATA_SIZE) to cover all cascade + XTS keys.
    /// </summary>
    public static byte[] DeriveKey(byte[] password, byte[] salt, TrueCryptPrf prf, bool boot)
    {
        int iterations = GetIterations(prf, boot);
        int outputLen = TrueCryptConstants.MasterKeyDataSize; // 256 bytes

        IDigest digest = prf switch
        {
            TrueCryptPrf.RipeMd160 => new RipeMD160Digest(),
            TrueCryptPrf.Sha512 => new Sha512Digest(),
            TrueCryptPrf.Whirlpool => new WhirlpoolDigest(),
            TrueCryptPrf.Sha1 => new Sha1Digest(),
            _ => throw new ArgumentException($"Unknown PRF: {prf}")
        };

        var generator = new Pkcs5S2ParametersGenerator(digest);
        generator.Init(password, salt, iterations);
        var keyParam = (KeyParameter)generator.GenerateDerivedMacParameters(outputLen * 8);
        return keyParam.GetKey();
    }
}

File: tc-app/src-tauri/src/core/key_derivation.rs
/// PBKDF2 key derivation supporting all TrueCrypt PRFs.

use crate::core::constants::MASTER_KEY_DATA_SIZE;
use hmac::Hmac;
use pbkdf2::pbkdf2_hmac;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Prf {
    RipeMd160,
    Sha512,
    Whirlpool,
    Sha1,
}

impl Prf {
    pub const ALL: &[Prf] = &[Prf::RipeMd160, Prf::Sha512, Prf::Whirlpool, Prf::Sha1];

    pub fn name(&self) -> &'static str {
        match self {
            Prf::RipeMd160 => "HMAC-RIPEMD-160",
            Prf::Sha512 => "HMAC-SHA-512",
            Prf::Whirlpool => "HMAC-Whirlpool",
            Prf::Sha1 => "HMAC-SHA-1",
        }
    }
}

/// Returns the PBKDF2 iteration count for a given PRF.
pub fn get_iterations(prf: Prf, boot: bool) -> u32 {
    match prf {
        Prf::RipeMd160 => if boot { 1000 } else { 2000 },
        Prf::Sha512 => 1000,
        Prf::Whirlpool => 1000,
        Prf::Sha1 => 2000,
    }
}

/// Derives a key using PBKDF2-HMAC with the specified PRF.
/// Output is MASTER_KEY_DATA_SIZE (256) bytes.
pub fn derive_key(
    password: &[u8],
    salt: &[u8],
    prf: Prf,
    boot: bool,
) -> [u8; MASTER_KEY_DATA_SIZE] {
    let iterations = get_iterations(prf, boot);
    let mut output = [0u8; MASTER_KEY_DATA_SIZE];

    match prf {
        Prf::RipeMd160 => {
            pbkdf2_hmac::<ripemd::Ripemd160>(password, salt, iterations, &mut output);
        }
        Prf::Sha512 => {
            pbkdf2_hmac::<sha2::Sha512>(password, salt, iterations, &mut output);
        }
        Prf::Whirlpool => {
            // whirlpool crate implements the Digest trait; use hmac manually
            pbkdf2::pbkdf2::<Hmac<whirlpool::Whirlpool>>(password, salt, iterations, &mut output)
                .expect("HMAC can be initialized with any key length");
        }
        Prf::Sha1 => {
            pbkdf2_hmac::<sha1::Sha1>(password, salt, iterations, &mut output);
        }
    }

    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_derive_key_sha512_produces_output() {
        let password = b"test_password";
        let salt = [0u8; 64];
        let key = derive_key(password, &salt, Prf::Sha512, false);
        // Just verify it's not all zeros (actual correctness verified against C# app)
        assert!(key.iter().any(|&b| b != 0));
    }

    #[test]
    fn test_all_prfs_produce_output() {
        let password = b"hello";
        let salt = [0x42u8; 64];
        for &prf in Prf::ALL {
            let key = derive_key(password, &salt, prf, false);
            assert!(key.iter().any(|&b| b != 0), "PRF {:?} produced all zeros", prf);
        }
    }
}
